### DISCUSSION TITLE
Typescript type for Realm.Results

### DISCUSSION THREAD

**Discussion by AudreyKj** [realm/realm-js#3921] [Q&A]
Hello! I'm using Realm JS for a React-Native app written in TypeScript and I'm wondering about the typing of `Realm.Results`. 

here is my use case: 

My Schema for 'MessageData' is: 

```
export const MessageDataSchema = {
  name: 'MessageData',
  primaryKey: 'id',
  properties: {
  id: 'string',
  messages: { type: 'list', objectType: 'Message' },
  }
}
```
Here is the typing is use to fetch data from the database (fetches the MessageData of a particular conversation, with its conversationId).

` const databaseMessages: Realm.Results<MessageData> = realm.objectForPrimaryKey<Realm.Results<MessageData>>('MessageData', conversationId)`

With this type, I get an error further down my code when I add a listener to the database: 

Property 'messages' does not exist on type 'Results<MessageData>'.ts(2339)

```
  databaseMessages.addListener(() => {
        setMessages([...databaseMessages.messages]);
      });
```

any thoughts why this might be happening? thanks a lot!

**Comment by SMhdAsadi** [comment]
I have some suggestions to make your Typescript experience better.

1. Use one class per schema as described [here](https://www.mongodb.com/docs/atlas/device-sdks/sdk/react-native/model-data/define-a-realm-object-model/). Check out Typescript sections. It helps Typescript infer the object types better.
2. Use Realm React package. With the hooks introduced in this package you usually don't need to deal with `realm` object. It just feels better in React.
3. Even if you have to use `objectForPrimaryKey`, there is no need to give it the exact types. It should just work and infer the types correctly. You need to give it your schema class described in section 1 as first argument.

Hope this helps.
